### ISSUE TITLE
PHP fatal error when lazyloading images

### ISSUE THREAD

**Issue by vmanthos** [wp-media/wp-rocket#4767]
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version ✅ `3.10.8`
- Used the search feature to ensure that the bug hasn’t been reported before ✅

**Describe the bug**

A PHP fatal error, like the following:
```
PHP Fatal error: Uncaught TypeError: preg_match_all() expects parameter 2 to be string, null given in /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php:257

Stack trace:
#0 /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php(257): preg_match_all('#<picture(?:.*)...', NULL, NULL, 2)
#1 /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(372): WP_Rocket\\Dependencies\\RocketLazyload\\Image->lazyloadPictures('<!DOCTYPE html>...', NULL)
#2 /nas/content/live/exampleprod/wp-includes/class-wp-hook.php(303): WP_Rocket\\Engine\\Media\\Lazyload\\Subscriber->lazyload('<!DOCTYPE html>...')
#3 /nas/content/live/exampleprod/wp-includes/plugin.php(189): WP_Hook->apply_filters('<!DOCTYPE html>...', Array)
#4 /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-optimization.php(104): apply_filters('rocket_buffer', '<!DOCTYPE html>...')\
n#5 [internal function","PHP Fatal error: Uncaught TypeError: preg_match_all() expects parameter 2 to be string, null given in /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php:257
Stack trace:
#0 /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php(257): preg_match_all('#<picture(?:.*)...', NULL, NULL, 2)
#1 /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(372): WP_Rocket\\Dependencies\\RocketLazyload\\Image->lazyloadPictures('<!DOCTYPE html>...', NULL)
#2 /nas/content/live/exampleprod/wp-includes/class-wp-hook.php(303): WP_Rocket\\Engine\\Media\\Lazyload\\Subscriber->lazyload('<!DOCTYPE html>...')
#3 /nas/content/live/exampleprod/wp-includes/plugin.php(189): WP_Hook->apply_filters('<!DOCTYPE html>...', Array)
#4 /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-optimization.php(104): apply_filters('rocket_buffer', '<!DOCTYPE html>...')
#5 [internal function]: WP_Rocket\\Buffe in /nas/content/live/exampleprod/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php on line 257, referer: https://www.example.com/collaboration/","Fri Feb 25 2022 11:31:34 GMT+0530 (India Standard Time)"
```

occurs if either of the following two lines:
https://github.com/wp-media/wp-rocket/blob/c2f41672833937b3025e6b6acf59cf9f27116007/inc/Engine/Media/Lazyload/Subscriber.php#L359-L360

returns `NULL`, e.g. because of a `PREG_BACKTRACK_LIMIT_ERROR`.

This is because we aren't safeguarding the `ignore_scripts()` and `ignore_noscripts()` methods from returning such a value.

**To Reproduce**

The page where this occurred can be downloaded from here:
https://drive.google.com/uc?id=18G-DrCTU1ENDbDZndG79ACHXk5-Wgckd

**Expected behavior**

No PHP fatal errors should occur with lazyload.

**Additional context**

We can check what both methods return, and if `empty()` return the unprocessed HTML.
The feature will not be applied, but no PHP fatal errors will occur.

Also, given that we apply those methods in the subscriber, I'm thinking we don't need these two lines:
https://github.com/wp-media/wp-rocket/blob/b34218065e5935e7482c7c60721becc26e237c97/inc/Dependencies/RocketLazyload/Image.php#L26-L27

If we need them, we have to safeguard them as well.

**Ticket:** https://secure.helpscout.net/conversation/1797267220/328111?folderId=2135277
The customer didn't provide access, so the investigation was done only by checking the code.

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1803266318/329150?folderId=2135277

From the `debug.log`:
```
[03-Mar-2022 10:46:53 UTC] PHP Fatal error:  Uncaught TypeError: preg_replace(): Argument #3 ($subject) must be of type array|string, null given in /var/www/vhosts/example.com/httpdocs/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php:27
Stack trace:
#0 /var/www/vhosts/example.com/httpdocs/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php(27): preg_replace()
#1 /var/www/vhosts/example.com/httpdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(375): WP_Rocket\Dependencies\RocketLazyload\Image->lazyloadImages()
#2 /var/www/vhosts/example.com/httpdocs/wp-includes/class-wp-hook.php(303): WP_Rocket\Engine\Media\Lazyload\Subscriber->lazyload()
#3 /var/www/vhosts/example.com/httpdocs/wp-includes/plugin.php(189): WP_Hook->apply_filters()
#4 /var/www/vhosts/example.com/httpdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-optimization.php(104): apply_filters()
#5 [internal function]: WP_Rocket\Buffer\Optimization->maybe_process_buffer()
#6 /var/www/vhosts/example.com/httpdocs/wp-includes/functions.php(5107): ob_end_flush()
#7 /var/www/vhosts/example.com/httpdocs/wp-includes/class-wp-hook.php(303): wp_ob_end_flush_all()
#8 /var/www/vhosts/example.com/httpdocs/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters()
#9 /var/www/vhosts/example.com/httpdocs/wp-includes/plugin.php(470): WP_Hook->do_action()
#10 /var/www/vhosts/example.com/httpdocs/wp-includes/load.php(1097): do_action()
#11 [internal function]: shutdown_action_hook()
#12 {main}
  thrown in /var/www/vhosts/example.com/httpdocs/wp-content/plugins/wp-rocket/inc/Dependencies/RocketLazyload/Image.php on line 27
```
In this case, adding `ini_set('pcre.backtrack_limit', PHP_INT_MAX);` in the `wp-config.php` file, stopped the fatal error and allowed lazyload to be applied.

**Comment by remyperona** [comment]
### Scope a solution ✅ 
We should update this along #4759, to use the `RegexTrait` too.

**Comment by CrochetFeve0251** [comment]
I added  ignore_scripts() and ignore_noscripts() in the RegexTrait in the PR from #4759, I will fix these issue when PR from #4759 is merged

**Comment by remyperona** [comment]
@CrochetFeve0251 You can update this class in the same PR
